File: apps/web/src/api/modelProviderConnectionRequirement.ts
export const MODEL_PROVIDER_CONNECTION_REQUIRED_EVENT = 'agentweaver:model-provider-connection-required';
export const MODEL_PROVIDER_CONNECTION_REQUIRED_CODE = 'model_provider_connection_required';
export const MODEL_PROVIDER_CONNECTION_REQUIRED_MESSAGE =
  'Connect your GitHub Copilot account to continue.';
export const CONFIGURE_PROJECT_MODEL_PROVIDER_ACTION = 'configure_project_model_provider';
export const CONFIGURE_PLATFORM_MODEL_PROVIDER_ACTION = 'configure_platform_model_provider';

export interface ModelProviderConnectionRequirement {
  code: string;
  message: string;
  action: {
    type: typeof CONFIGURE_PROJECT_MODEL_PROVIDER_ACTION | typeof CONFIGURE_PLATFORM_MODEL_PROVIDER_ACTION;
    project_id: string;
  };
}

export function isModelProviderConnectionRequirement(value: unknown): value is ModelProviderConnectionRequirement {
  if (typeof value !== 'object' || value === null) return false;
  const record = value as Record<string, unknown>;
  const action = record.action;
  if (typeof record.code !== 'string' || typeof record.message !== 'string') return false;
  if (typeof action !== 'object' || action === null) return false;
  const actionType = (action as Record<string, unknown>).type;
  return (actionType === CONFIGURE_PROJECT_MODEL_PROVIDER_ACTION || actionType === CONFIGURE_PLATFORM_MODEL_PROVIDER_ACTION)
    && typeof (action as Record<string, unknown>).project_id === 'string';
}

/** True when a requirement is scoped to a specific project, rather than the platform default. */
export function isProjectScopedModelProviderRequirement(requirement: ModelProviderConnectionRequirement): boolean {
  return requirement.action.type === CONFIGURE_PROJECT_MODEL_PROVIDER_ACTION;
}

File: apps/Agentweaver.Api/Auth/ModelProviderConnectionRequirement.cs
using System.Text.Json.Serialization;
using Agentweaver.AgentRuntime.Providers;
using Agentweaver.Domain;

namespace Agentweaver.Api.Auth;

/// <summary>
/// The single redacted response contract for a model-provider capability that must be connected
/// by a human. Its action deliberately identifies either an existing project-scoped handoff or
/// the platform-settings handoff rather than carrying a credential, OAuth URL, transaction, or
/// callback state.
/// </summary>
public sealed record ModelProviderConnectionRequirement(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("action")] ModelProviderConnectionAction Action)
{
    public const string RequirementCode = "model_provider_connection_required";
    public const string RequirementMessage = ProjectRequirementMessage;
    public const string ProjectRequirementMessage =
        "Connect the project's GitHub Copilot App to continue.";
    public const string PlatformDefaultRequirementMessage =
        "Connect the platform-default GitHub Copilot account to continue.";

    public static ModelProviderConnectionRequirement ForProject(ProjectId projectId) =>
        new(
            RequirementCode,
            ProjectRequirementMessage,
            new ModelProviderConnectionAction(
                ModelProviderConnectionAction.ConfigureProjectModelProvider,
                projectId.ToString()));

    public static ModelProviderConnectionRequirement ForPlatformDefault() =>
        new(
            RequirementCode,
            PlatformDefaultRequirementMessage,
            new ModelProviderConnectionAction(
                ModelProviderConnectionAction.ConfigurePlatformModelProvider,
                string.Empty));
}

/// <summary>
/// Raised before a run can enter an AgentHost launch path when its project-scoped or platform-wide
/// model-provider capability is not redeemable. Endpoint surfaces translate this to the single
/// redacted connection-required response contract.
/// </summary>
public sealed class ModelProviderConnectionRequiredException : AgentProviderException
{
    public ModelProviderConnectionRequiredException(ProjectId projectId)
        : this(ModelProviderConnectionRequirement.ForProject(projectId))
    {
    }

    public ModelProviderConnectionRequiredException()
        : this(ModelProviderConnectionRequirement.ForPlatformDefault())
    {
    }

    private ModelProviderConnectionRequiredException(ModelProviderConnectionRequirement requirement)
        : base(
            ModelSource.GitHubCopilot,
            AgentProviderFailureKind.Authorization,
            requirement.Code,
            requirement.Message,
            isRetryable: false)
    {
        Requirement = requirement;
    }

    public ModelProviderConnectionRequirement Requirement { get; }
}

/// <summary>
/// Typed action consumed by every UI surface that receives a
/// <see cref="ModelProviderConnectionRequirement"/>. The action type distinguishes project scope
/// from platform scope directly (rather than via an empty-string project id sentinel), so the
/// client can route a project-scoped requirement to the project's model-provider settings and a
/// platform-scoped requirement to Platform Settings so a Platform Admin can connect the
/// deployment-wide account.
/// </summary>
public sealed record ModelProviderConnectionAction(
    [property: JsonPropertyName("type")] string Type,
    [property: JsonPropertyName("project_id")] string ProjectId)
{
    public const string ConfigureProjectModelProvider = "configure_project_model_provider";
    public const string ConfigurePlatformModelProvider = "configure_platform_model_provider";
}
